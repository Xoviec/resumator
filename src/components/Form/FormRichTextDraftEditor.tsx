import FormatBoldIcon from "@mui/icons-material/FormatBold";
import FormatItalicIcon from "@mui/icons-material/FormatItalic";
// Icons
import FormatListBulletedIcon from "@mui/icons-material/FormatListBulleted";
import FormatListNumberedIcon from "@mui/icons-material/FormatListNumbered";
import FormatUnderlinedIcon from "@mui/icons-material/FormatUnderlined";
import { Box, Divider } from "@mui/material";
import {
  ContentState,
  convertFromHTML,
  convertFromRaw,
  convertToRaw,
  Editor,
  EditorState,
  RichUtils,
} from "draft-js";
import "draft-js/dist/Draft.css";
import {
  MouseEvent,
  ReactNode,
  useRef,
  useState,
  VoidFunctionComponent,
} from "react";
import { TooltipIconButton } from "../Material";

interface FormRichTextDraftEditorProps {
  value: string;
  onChange: (value: string) => void;
}

export const FormRichTextDraftEditor: VoidFunctionComponent<
  FormRichTextDraftEditorProps
> = ({ value, onChange }) => {
  let content;

  // Setup the editor.
  try {
    content = convertFromRaw(JSON.parse(value));
  } catch (err) {
    // If parsing fails, treat it as a raw text string.
    const { contentBlocks, entityMap } = convertFromHTML(value);
    content = ContentState.createFromBlockArray(contentBlocks, entityMap);
  }

  const editorRef = useRef<Editor>(null);
  const [editorState, setEditorState] = useState(
    EditorState.createWithContent(content)
  );

  const handleFocus = (event: MouseEvent) => {
    event.preventDefault();
    event.stopPropagation();
    editorRef?.current?.focus();
  };

  /**
   * Save the changes to the form when focus is lost.
   */
  const handleBlur = () => {
    const raw = convertToRaw(editorState.getCurrentContent());
    onChange(JSON.stringify(raw));
  };

  const renderBlockTypeButton = (type: string, tooltip: string, icon: ReactNode) => (
    <TooltipIconButton
      data-testid={`${type}-block-type-button`}
      tooltip={tooltip}
      active={hasBlockType(type)}
      onClick={() => toggleBlockType(type)}
      onMouseDown={(event) => event.preventDefault()}
    >
      {icon}
    </TooltipIconButton>
  );

  /**
   * Set a block type on the current cursor location or selected text.
   */
  const toggleBlockType = (type: string) => {
    setEditorState(RichUtils.toggleBlockType(editorState, type));
  };

  const hasBlockType = (type: string) => {
    const selection = editorState.getSelection();
    const block = editorState
      .getCurrentContent()
      .getBlockForKey(selection.getStartKey());
    return block.getType() === type;
  };

  const renderStyleButton = (style: string, tooltip: string, icon: ReactNode) => (
    <TooltipIconButton
      data-testid={`${style}-style-button`}
      tooltip={tooltip}
      active={hasStyle(style)}
      onClick={() => toggleStyle(style)}
      onMouseDown={(event) => event.preventDefault()}
    >
      {icon}
    </TooltipIconButton>
  );

  /**
   * Set a style on the current cursor location or selected text.
   */
  const toggleStyle = (style: string) => {
    setEditorState(RichUtils.toggleInlineStyle(editorState, style));
  };

  const hasStyle = (style: string) => editorState.getCurrentInlineStyle().has(style);

  return (
    <Box
      data-testid="form-rich-text-draft-editor-container"
      onClick={handleFocus}
      onBlur={handleBlur}
      sx={{
        "& .DraftEditor-root": {
          padding: "4.5px",
          height: "unset",
          overflow: "auto",
        },
        "& ul, ol": {
          margin: 0,
        },
        display: "flex",
        flexDirection: "column",
        gridGap: 4,
        paddingX: 1,
        paddingY: 1,
        height: 400,
        flex: 1,
      }}
    >
      {/* Styling options */}
      <Box
        sx={{
          display: "flex",
          flexDirection: "row",
          alignItems: "center",
        }}
      >
        {/* Unordered list */}
        {renderBlockTypeButton(
          "unordered-list-item",
          "Unordered List",
          <FormatListBulletedIcon fontSize="small" />
        )}
        {/* Ordered list */}
        {renderBlockTypeButton(
          "ordered-list-item",
          "Ordered List",
          <FormatListNumberedIcon fontSize="small" />
        )}
        <Divider
          orientation="vertical"
          sx={{
            margin: "0 4px",
            height: "80%",
          }}
        />
        {/* Bold */}
        {renderStyleButton("BOLD", "Bold", <FormatBoldIcon fontSize="small" />)}
        {/* Italic */}
        {renderStyleButton(
          "ITALIC",
          "Italic",
          <FormatItalicIcon fontSize="small" />
        )}
        {/* Underlined */}
        {renderStyleButton(
          "UNDERLINE",
          "Underlined",
          <FormatUnderlinedIcon fontSize="small" />
        )}
      </Box>
      <Divider />
      <Editor
        spellCheck
        ref={editorRef}
        onTab={(e) => {
          e.preventDefault();
          setEditorState(RichUtils.onTab(e, editorState, 4));
        }}
        editorState={editorState}
        onChange={(state) => setEditorState(state)}
      />
    </Box>
  );
};
