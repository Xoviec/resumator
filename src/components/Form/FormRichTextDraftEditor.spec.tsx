import { render, fireEvent } from "@testing-library/react";
import { FormRichTextDraftEditor } from "./FormRichTextDraftEditor";
import * as draftJs from "draft-js";
import { mocked } from "jest-mock";

jest.mock("draft-js", () => {
  const reactLib = jest.requireActual("react");

  // unfortunately once we have a component with forwardRef, we can't spy on it or mock it, so workaround
  const editorPropReceiver = jest.fn();
  const Editor = reactLib.forwardRef((props: any, ref: any) => {
    editorPropReceiver(props);
    return (
      <input
        ref={ref}
        data-testid="editor-input"
        onChange={(e) => {
          props.onChange({
            ...props.editorState,
            id: `input-change-${e.target.value}`,
            getCurrentContent: () => ({
              inputValue: e.target.value,
              getBlockForKey: jest.fn(() => ({
                getType: jest.fn(() => "test"),
              })),
            }),
          });
        }}
      />
    );
  });
  const draftJsMock: any = jest.createMockFromModule("draft-js");

  return {
    ...draftJsMock,
    editorPropReceiver,
    Editor,
  };
});

const {
  EditorState,
  convertFromRaw,
  convertToRaw,
  convertFromHTML,
  RichUtils,
  ContentState,
} = draftJs;
const editorPropReceiver = mocked(
  // @ts-ignore
  draftJs.editorPropReceiver
);

const testStyleButton = (styleType: string) => {
  mocked(RichUtils.toggleInlineStyle).mockImplementation(
    (editorState) =>
      ({
        ...editorState,
        id: `${styleType}-set`,
        getCurrentContent: () => ({
          inputValue: `${styleType}-input-value`,
          getBlockForKey: jest.fn(() => ({
            getType: jest.fn(() => "test"),
          })),
        }),
      } as any)
  );

  const onChange = jest.fn();
  const { getByTestId } = render(
    <FormRichTextDraftEditor
      value={JSON.stringify({ test: "value" })}
      onChange={onChange}
    />
  );

  mocked(editorPropReceiver).mockClear();

  fireEvent.click(getByTestId(`${styleType}-style-button`));

  expect(mocked(RichUtils.toggleInlineStyle)).toHaveBeenCalledWith(
    expect.objectContaining({
      id: "default-editor-state",
    }),
    `${styleType}`
  );
  expect(mocked(editorPropReceiver)).toHaveBeenCalledWith(
    expect.objectContaining({
      editorState: expect.objectContaining({
        id: `${styleType}-set`,
      }),
    })
  );

  fireEvent.blur(getByTestId("form-rich-text-draft-editor-container"));
  expect(onChange).toHaveBeenCalledWith(
    JSON.stringify({ value: `${styleType}-input-value` })
  );
};

const testBlockTypeButton = (blockType: string) => {
  mocked(RichUtils.toggleBlockType).mockImplementation(
    (editorState) =>
      ({
        ...editorState,
        id: `${blockType}-set`,
        getCurrentContent: () => ({
          inputValue: `${blockType}-input-value`,
          getBlockForKey: jest.fn(() => ({
            getType: jest.fn(() => "test"),
          })),
        }),
      } as any)
  );

  const onChange = jest.fn();
  const { getByTestId } = render(
    <FormRichTextDraftEditor
      value={JSON.stringify({ test: "value" })}
      onChange={onChange}
    />
  );

  mocked(editorPropReceiver).mockClear();

  fireEvent.click(getByTestId(`${blockType}-block-type-button`));

  expect(mocked(RichUtils.toggleBlockType)).toHaveBeenCalledWith(
    expect.objectContaining({
      id: "default-editor-state",
    }),
    `${blockType}`
  );
  expect(mocked(editorPropReceiver)).toHaveBeenCalledWith(
    expect.objectContaining({
      editorState: expect.objectContaining({
        id: `${blockType}-set`,
      }),
    })
  );

  fireEvent.blur(getByTestId("form-rich-text-draft-editor-container"));
  expect(onChange).toHaveBeenCalledWith(
    JSON.stringify({ value: `${blockType}-input-value` })
  );
};

describe("FormRichTextDraftEditor", () => {
  const getDefaultEditorState = () =>
    ({
      id: "default-editor-state",
      getSelection: jest.fn(() => ({
        getStartKey: jest.fn(() => "test"),
      })),
      getCurrentContent: jest.fn(() => ({
        getBlockForKey: jest.fn(() => ({
          getType: jest.fn(() => "test"),
        })),
      })),
      getCurrentInlineStyle: jest.fn(() => ({
        has: jest.fn(() => true),
      })),
    } as any);

  beforeEach(() => {
    jest.resetAllMocks();

    mocked(convertFromRaw).mockImplementation(() => "default-test-content" as any);

    const convertToRawImplementation: any = (currentContent: any) => {
      if (currentContent?.inputValue) {
        return { value: currentContent?.inputValue };
      }

      return { value: "default-test-content" };
    };
    mocked(convertToRaw).mockImplementation(convertToRawImplementation);
    mocked(EditorState.createWithContent).mockImplementation(getDefaultEditorState);
  });

  it("should render correctly", () => {
    const { asFragment } = render(
      <FormRichTextDraftEditor
        value={JSON.stringify({ test: "value" })}
        onChange={() => {}}
      />
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it("should render the Editor with given value", () => {
    render(
      <FormRichTextDraftEditor
        value={JSON.stringify({ test: "value" })}
        onChange={() => {}}
      />
    );

    expect(mocked(convertFromRaw)).toHaveBeenCalledWith({ test: "value" });
    expect(mocked(EditorState.createWithContent)).toHaveBeenCalledWith(
      "default-test-content"
    );
    expect(editorPropReceiver).toHaveBeenCalledWith(
      expect.objectContaining({
        editorState: expect.objectContaining({
          id: "default-editor-state",
        }),
      })
    );
  });

  it("should fallback to the convertFromHTML when JSON parse fails", () => {
    mocked(convertFromHTML).mockImplementation(
      () =>
        ({
          contentBlocks: "content-blocks",
          entityMap: "entity-map",
        } as any)
    );
    mocked(ContentState.createFromBlockArray).mockImplementation(
      () => "default-test-content" as any
    );

    render(
      <FormRichTextDraftEditor value={"unparsable-value"} onChange={() => {}} />
    );

    expect(mocked(convertFromHTML)).toHaveBeenCalledWith("unparsable-value");
    expect(mocked(ContentState.createFromBlockArray)).toHaveBeenCalledWith(
      "content-blocks",
      "entity-map"
    );
    expect(editorPropReceiver).toHaveBeenCalledWith(
      expect.objectContaining({
        editorState: expect.objectContaining({
          id: "default-editor-state",
        }),
      })
    );
  });

  it("should forward focus to the Editor", () => {
    const { getByTestId } = render(
      <FormRichTextDraftEditor
        value={JSON.stringify({ test: "value" })}
        onChange={() => {}}
      />
    );

    fireEvent.click(getByTestId("form-rich-text-draft-editor-container"));
    expect(getByTestId("editor-input")).toHaveFocus();
  });

  it("should call onChange with correct values on blur", () => {
    const onChange = jest.fn();
    const { getByTestId } = render(
      <FormRichTextDraftEditor
        value={JSON.stringify({ test: "value" })}
        onChange={onChange}
      />
    );

    fireEvent.change(getByTestId("editor-input"), {
      target: { value: "test-value-from-input-event" },
    });
    fireEvent.blur(getByTestId("form-rich-text-draft-editor-container"));

    expect(onChange).toHaveBeenCalledWith(
      JSON.stringify({ value: "test-value-from-input-event" })
    );
  });

  it.each(["BOLD", "ITALIC", "UNDERLINE"])(
    "should have correct text when %s style button is clicked",
    (styleType: string) => {
      testStyleButton(styleType);
    }
  );

  it.each(["unordered-list-item", "ordered-list-item"])(
    "should have correct text when %s block type button is clicked",
    (blockType: string) => {
      testBlockTypeButton(blockType);
    }
  );
});
