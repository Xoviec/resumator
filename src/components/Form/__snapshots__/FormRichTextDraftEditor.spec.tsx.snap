// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormRichTextDraftEditor should render correctly 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root css-3zej7z"
    data-testid="form-rich-text-draft-editor-container"
  >
    <div
      class="MuiBox-root css-u4p24i"
    >
      <button
        aria-label="Unordered List"
        class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeLarge css-mf1cb5-MuiButtonBase-root-MuiIconButton-root"
        data-mui-internal-clone-element="true"
        data-testid="unordered-list-item-block-type-button"
        tabindex="0"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-ptiqhd-MuiSvgIcon-root"
          data-testid="FormatListBulletedIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M4 10.5c-.83 0-1.5.67-1.5 1.5s.67 1.5 1.5 1.5 1.5-.67 1.5-1.5-.67-1.5-1.5-1.5zm0-6c-.83 0-1.5.67-1.5 1.5S3.17 7.5 4 7.5 5.5 6.83 5.5 6 4.83 4.5 4 4.5zm0 12c-.83 0-1.5.68-1.5 1.5s.68 1.5 1.5 1.5 1.5-.68 1.5-1.5-.67-1.5-1.5-1.5zM7 19h14v-2H7v2zm0-6h14v-2H7v2zm0-8v2h14V5H7z"
          />
        </svg>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
      <button
        aria-label="Ordered List"
        class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeLarge css-mf1cb5-MuiButtonBase-root-MuiIconButton-root"
        data-mui-internal-clone-element="true"
        data-testid="ordered-list-item-block-type-button"
        tabindex="0"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-ptiqhd-MuiSvgIcon-root"
          data-testid="FormatListNumberedIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M2 17h2v.5H3v1h1v.5H2v1h3v-4H2v1zm1-9h1V4H2v1h1v3zm-1 3h1.8L2 13.1v.9h3v-1H3.2L5 10.9V10H2v1zm5-6v2h14V5H7zm0 14h14v-2H7v2zm0-6h14v-2H7v2z"
          />
        </svg>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
      <hr
        class="MuiDivider-root MuiDivider-fullWidth MuiDivider-vertical css-pjtdqm-MuiDivider-root"
      />
      <button
        aria-label="Bold"
        class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorSecondary MuiIconButton-sizeLarge css-1f88mc5-MuiButtonBase-root-MuiIconButton-root"
        data-mui-internal-clone-element="true"
        data-testid="BOLD-style-button"
        tabindex="0"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-ptiqhd-MuiSvgIcon-root"
          data-testid="FormatBoldIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M15.6 10.79c.97-.67 1.65-1.77 1.65-2.79 0-2.26-1.75-4-4-4H7v14h7.04c2.09 0 3.71-1.7 3.71-3.79 0-1.52-.86-2.82-2.15-3.42zM10 6.5h3c.83 0 1.5.67 1.5 1.5s-.67 1.5-1.5 1.5h-3v-3zm3.5 9H10v-3h3.5c.83 0 1.5.67 1.5 1.5s-.67 1.5-1.5 1.5z"
          />
        </svg>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
      <button
        aria-label="Italic"
        class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorSecondary MuiIconButton-sizeLarge css-1f88mc5-MuiButtonBase-root-MuiIconButton-root"
        data-mui-internal-clone-element="true"
        data-testid="ITALIC-style-button"
        tabindex="0"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-ptiqhd-MuiSvgIcon-root"
          data-testid="FormatItalicIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M10 4v3h2.21l-3.42 8H6v3h8v-3h-2.21l3.42-8H18V4z"
          />
        </svg>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
      <button
        aria-label="Underlined"
        class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorSecondary MuiIconButton-sizeLarge css-1f88mc5-MuiButtonBase-root-MuiIconButton-root"
        data-mui-internal-clone-element="true"
        data-testid="UNDERLINE-style-button"
        tabindex="0"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-ptiqhd-MuiSvgIcon-root"
          data-testid="FormatUnderlinedIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M12 17c3.31 0 6-2.69 6-6V3h-2.5v8c0 1.93-1.57 3.5-3.5 3.5S8.5 12.93 8.5 11V3H6v8c0 3.31 2.69 6 6 6zm-7 2v2h14v-2H5z"
          />
        </svg>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
    <hr
      class="MuiDivider-root MuiDivider-fullWidth css-9mgopn-MuiDivider-root"
    />
    <input
      data-testid="editor-input"
    />
  </div>
</DocumentFragment>
`;
